### PR TITLE
Implement `arange` (plus a grab bag of smaller things)

### PR DESCRIPTION
**Pull request by inducer**
- [x] `arange` needs tests

Probably best to read commit-by-commit.

cc @isuruf

### CODE REVIEW

File: pytato/array.py [hunk 1]
```diff
@@ -2025,6 +2027,91 @@ def eye(N: int, M: Optional[int] = None, k: int = 0,  # noqa: N803
...
+        if k in admissible_names:
+            if getattr(inf, k) is None:
+                setattr(inf, k, v)
+                if k == "dtype":
+                    explicit_dtype = True
+            else:
+                raise ValueError(
+                        "may not specify '%s' by position and keyword" % k)
+        else:
+            raise ValueError("unexpected keyword argument '%s'" % k)
```
Comment (kaushikcfd) [inline pytato/array.py:2086]:
```suggestion
                raise TypeError(
                        "may not specify '%s' by position and keyword" % k)
        else:
            raise TypeError("unexpected keyword argument '%s'" % k)
```

File: pytato/array.py [hunk 2]
```diff
@@ -2025,6 +2027,91 @@ def eye(N: int, M: Optional[int] = None, k: int = 0,  # noqa: N803
...
+        inf.stop = args[0]
+    elif argc == 2:
+        inf.start = args[0]
+        inf.stop = args[1]
+    elif argc == 3:
+        inf.start = args[0]
+        inf.stop = args[1]
+        inf.step = args[2]
+    else:
+        raise ValueError("too many arguments")
```
Comment (kaushikcfd) [inline pytato/array.py:2073]:
```suggestion
        raise TypeError("arange() takes 0 to 4 positional arguments but"
                        f" {argc} were given")
```
NumPy raises a type error.

File: pytato/array.py [hunk 3]
```diff
@@ -2025,6 +2027,91 @@ def eye(N: int, M: Optional[int] = None, k: int = 0,  # noqa: N803
...
+    if not explicit_dtype:
+        raise TypeError("arange requires a dtype argument")
+
+    dtype = np.dtype(inf.dtype)
+    start = dtype.type(inf.start)
+    step = dtype.type(inf.step)
+    stop = dtype.type(inf.stop)
+
+    from math import ceil
+    size = max(0, int(ceil((stop-start)/step)))
```
Comment (kaushikcfd) [inline pytato/array.py:2106]: raise a `NotImpleementedError` for SizeParam `start,stop`?

**Review by kaushikcfd — APPROVED** [review]
LGTM, thanks!

File: pytato/array.py (hunk 3 above)
Comment (inducer) [reply pytato/array.py:2117]: https://github.com/inducer/pytato/compare/88fb1da90f31c62c97a9da4474437ef1d4fcb2d5..01da3534d17727b3150637291a46565f7ecc8dee

File: pytato/array.py (hunk 2 above)
Comment (inducer) [reply pytato/array.py]: https://github.com/inducer/pytato/compare/88fb1da90f31c62c97a9da4474437ef1d4fcb2d5..01da3534d17727b3150637291a46565f7ecc8dee

File: pytato/array.py (hunk 1 above)
Comment (inducer) [reply pytato/array.py]: https://github.com/inducer/pytato/compare/88fb1da90f31c62c97a9da4474437ef1d4fcb2d5..01da3534d17727b3150637291a46565f7ecc8dee

**Comment by inducer** [comment]
Thanks for taking a look!

**Pull request merged**